### ISSUE TITLE
Linux: 100% CPU usage when stdout is redirected to file/pipe/FIFO

### ISSUE THREAD

**Issue by powerman** [yar229/WebDavMailRuCloud#57]
`mono wdmrc.exe` doesn't use CPU when idle, but `mono wdmrc.exe | cat` and `mono wdmrc.exe >wdmrc.log` use 100% CPU all of time. Same happens when stdout is redirected to `/dev/null` and even when I run app with closed FD 1 (stdout).

Not sure is this mono issue or can be fixed in your code. Another `mono` app I use (KeePass) doesn't have this behaviour, but it also doesn't log anything to stdout. It doesn't call any syscalls while using 100% CPU (`strace mono wdmrc.exe` outputs nothing after initialization).

This application is a daemon (a service) by nature, so it's important to be able to run it in background, which means we must be able to redirect it logs to pipe, file or even `/dev/null`. (As temporary workaround you can add option to disable all output to stdout, this may helps a bit.)

**Comment by powerman** [comment]
Another workaround is run mono with allocated pseudo-tty using https://github.com/abligh/pty tool: `pty mono wdmrc.exe | cat` works just fine.

**Comment by yar229** [comment]
Edit `wdmrc.exe.config`, in section `<log4net>` add
```
<appender name="ConsoleAppender" type="log4net.Appender.ConsoleAppender" >
  <layout type="log4net.Layout.PatternLayout">
    <conversionPattern value="%date %-5level %logger - %message%newline" />
  </layout>
</appender>
<appender name="RemoteSyslogAppender" type="log4net.Appender.RemoteSyslogAppender">
  <remoteAddress value="127.0.0.1" />
  <layout type="log4net.Layout.PatternLayout, log4net">
    <conversionPattern value="%-5level %logger - %message%newline" />
  </layout>
</appender>
<appender name="LocalSyslogAppender" type="log4net.Appender.LocalSyslogAppender">
  <layout type="log4net.Layout.PatternLayout, log4net">
    <conversionPattern value="%-5level %logger - %message%newline" />
  </layout>
</appender>
```

in `<log4net><root>` instead of `<appender-ref ref="ColoredConsoleAppender"/>` similarly add ConsoleAppender (or RemoteSyslogAppender, or LocalSyslogAppender - if you want)

**Comment by powerman** [comment]
This won't helps. Log output is changes to non-colored or syslog, but even with syslog app still output 3 lines to stdout on start and in all cases when stdout is redirected to not-a-tty mono use 100% CPU.

**Comment by yar229** [comment]
How about not using redirect, but RollingFileAppender or
```
<appender name="FileAppender" type="log4net.Appender.FileAppender">
  <file value="debug.log" />
  <appendToFile value="false" />
  <layout type="log4net.Layout.PatternLayout">
    <conversionPattern value="%-8timestamp [%thread] %-5level %logger - %message%newline" />
  </layout>
</appender>
```

I'll try to understand what's wrog with redirection

**Comment by powerman** [comment]
All my system services use common way to save logs to files, filter them, rotate, etc. and thus using app-specific way to do same things is inconvenient. Also, just like with syslog, app still outputs 3 lines to stdout on start, which daemons shouldn't do.

**Comment by powerman** [comment]
Also, even with FileAppender (with `/dev/stdout` as file name) it still outputs few ANSI escape sequences (not suitable for log files):

```
0000  1b 5b 36 6e 1b 5b 48 1b  5b 4a ef bb bf 49 4e 46  .[6n.[H. [J...INF
0010  4f 20 20 59 61 52 2e 57  65 62 44 61 76 4d 61 69  O  YaR.W ebDavMai
0020  6c 52 75 2e 50 72 6f 67  72 61 6d 20 2d 20 57 65  lRu.Prog ram - We
0030  62 44 41 56 20 73 65 72  76 65 72 20 72 75 6e 6e  bDAV ser ver runn
0040  69 6e 67 20 61 74 20 68  74 74 70 3a 2f 2f 31 32  ing at h ttp://12
0050  37 2e 30 2e 30 2e 31 3a  31 30 38 30 31 0d 0a     7.0.0.1: 10801..
```

**Comment by yar229** [comment]
Ok, got it, I'll try to fix

**Comment by powerman** [comment]
And FileAppender also outputs BOM (Unicode byte order mark, `<0xEF,0xBB,0xBF>`) before first line of app output `WebDAVCloudMailRu: WebDAV emulator for Cloud.Mail.ru` and before first line of log output `[1]: WebDAV server running at`, which also isn't really suitable for logs (especially because common timestamp gets prepended to log lines, and BOM location became middle of line instead of start of file).

So, for logs we need some output which won't include anything except text - no BOM, no ANSI escapes, and works with redirected stdout/stderr.

**Comment by yar229** [comment]
Plz try [1.5.63](https://github.com/yar229/WebDavMailRuCloud/releases/tag/1.5.63)
Maybe the [same bug](https://github.com/yar229/WebDavMailRuCloud/issues/71)

**Comment by powerman** [comment]
I didn't forget about this, just a bit busy last days. I plan to check this in few days.

**Comment by AliakseiT** [comment]
Was the issue for me too. I can confirm it is solved in 1.5.63
